### PR TITLE
SONAR-10052 perf hotspot when synchronizing Quality profiles at startup

### PR DESCRIPTION
**Pull request by simonbrandhof**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileTreeImpl.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    Collection<RuleKey> ruleKeys = parentActiveRules.stream().map(ActiveRuleDto::getRuleKey).collect(MoreCollectors.toArrayList());
+    RuleActivationContext context = ruleActivator.createContextForUserProfile(dbSession, profile, ruleKeys);
+
+    for (ActiveRuleDto parentActiveRule : parentActiveRules) {
+      try {
+        RuleActivation activation = RuleActivation.create(parentActiveRule.getRuleKey(), null, null);
+        changes.addAll(ruleActivator.activate(dbSession, activation, context));
+      } catch (BadRequestException e) {
+        // for example because rule status is REMOVED
+        // TODO return errors
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileTreeImpl.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsb%2FSONAR-10052%2Fqprofiles&issues=AWE-qvPmmckxG3vIth73&open=AWE-qvPmmckxG3vIth73)

[ISSUE_KEY]: AWE-qvPmmckxG3vIth73

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java [hunk 2]
```diff
@@ -60,79 +62,65 @@
...
+    context.reset(activation.getRuleKey());
     return doActivate(dbSession, activation, context);
   }
 
-  public List<ActiveRuleChange> activateAndCommit(DbSession dbSession, RuleActivation activation, QProfileDto profile) {
-    List<ActiveRuleChange> changes = activate(dbSession, activation, profile);
-    activeRuleIndexer.commitAndIndex(dbSession, changes);
-    return changes;
-  }
+  private List<ActiveRuleChange> doActivate(DbSession dbSession, RuleActivation activation, RuleActivationContext context) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 30 to the 15 allowed. ([squid:S3776](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsb%2FSONAR-10052%2Fqprofiles&issues=AWE-qvP5mckxG3vIth78&open=AWE-qvP5mckxG3vIth78)

[ISSUE_KEY]: AWE-qvP5mckxG3vIth78

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java [hunk 3]
```diff
@@ -59,37 +60,37 @@
 
   private final DbClient dbClient;
   private final RuleFinder ruleFinder;
-  private final RuleActivator ruleActivator;
+  private final QProfileRules qProfileRules;
   private final ProfileExporter[] exporters;
   private final ProfileImporter[] importers;
 
-  public QProfileExporters(DbClient dbClient, RuleFinder ruleFinder, RuleActivator ruleActivator, ProfileExporter[] exporters, ProfileImporter[] importers) {
+  public QProfileExporters(DbClient dbClient, RuleFinder ruleFinder, QProfileRules qProfileRules, ProfileExporter[] exporters, ProfileImporter[] importers) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java]: Code Smell Code Smell: Remove this use of "RuleFinder"; it is deprecated. ([squid:CallToDeprecatedMethod](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsb%2FSONAR-10052%2Fqprofiles&issues=AWE-qvPNmckxG3vIth7y&open=AWE-qvPNmckxG3vIth7y)

[ISSUE_KEY]: AWE-qvPNmckxG3vIth7y

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java [hunk 4]
```diff
@@ -423,118 +363,119 @@ private String validateParam(RuleParamDto ruleParam, @Nullable String value) {
...
+    RuleActivationContext.Builder builder = new RuleActivationContext.Builder();
+
+    // load rules
+    List<RuleDefinitionDto> rules = completeWithRules(dbSession, builder, ruleKeys);
+
+    // load descendant profiles
+    List<QProfileDto> profiles = new ArrayList<>(db.qualityProfileDao().selectDescendants(dbSession, profile));
+    profiles.add(profile);
+    if (profile.getParentKee() != null) {
+      profiles.add(db.qualityProfileDao().selectByUuid(dbSession, profile.getParentKee()));
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java]: Bug Bug: "selectByUuid" is marked "javax.annotation.ParametersAreNonnullByDefault" but parameter 2 could be null. ([squid:S2637](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2637))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsb%2FSONAR-10052%2Fqprofiles&issues=AWE-qvP5mckxG3vIth75&open=AWE-qvP5mckxG3vIth75)

[ISSUE_KEY]: AWE-qvP5mckxG3vIth75

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java [hunk 5]
```diff
@@ -290,116 +272,74 @@ private ActiveRuleDto doInsert(ActiveRuleChange change, RuleActivatorContext con
...
+    }
+    activeRule.get().setUpdatedAt(system2.now());
+    dao.update(dbSession, activeRule.get());
+
+    for (Map.Entry<String, String> param : change.getParameters().entrySet()) {
+      ActiveRuleParamDto activeRuleParamDto = activeRule.getParam(param.getKey());
+      if (activeRuleParamDto == null) {
+        // did not exist
+        if (param.getValue() != null) {
+          activeRuleParamDto = ActiveRuleParamDto.createFor(context.getRule().getParam(param.getKey()));
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java]: Bug Bug: "createFor" is marked "javax.annotation.ParametersAreNonnullByDefault" but parameter 1 could be null. ([squid:S2637](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2637))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsb%2FSONAR-10052%2Fqprofiles&issues=AWE-qvP5mckxG3vIth76&open=AWE-qvP5mckxG3vIth76)

[ISSUE_KEY]: AWE-qvP5mckxG3vIth76

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java [hunk 6]
```diff
@@ -423,118 +363,119 @@ private String validateParam(RuleParamDto ruleParam, @Nullable String value) {
...
+    RuleActivationContext.Builder builder = new RuleActivationContext.Builder();
+
+    // load rules
+    List<RuleDefinitionDto> rules = completeWithRules(dbSession, builder, ruleKeys);
+
+    // load profiles
+    List<QProfileDto> aliasedBuiltInProfiles = db.qualityProfileDao().selectQProfilesByRuleProfile(dbSession, builtInProfile);
+    List<QProfileDto> profiles = new ArrayList<>(aliasedBuiltInProfiles);
+    for (QProfileDto aliasProfile : aliasedBuiltInProfiles) {
+      // FIXME N+1 syndrome. To be optimized by replacing db column parent_kee by ancestor_keys.
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java]: Code Smell Code Smell: Take the required action to fix the issue indicated by this comment. ([squid:S1134](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsb%2FSONAR-10052%2Fqprofiles&issues=AWE-qvP5mckxG3vIth77&open=AWE-qvP5mckxG3vIth77)

[ISSUE_KEY]: AWE-qvP5mckxG3vIth77

File: server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java [hunk 7]
```diff
@@ -130,6 +130,8 @@ public void start() {
       keysToIndex.addAll(removedRules.stream().map(RuleDefinitionDto::getKey).collect(Collectors.toList()));
 
       persistRepositories(dbSession, context.repositories());
+      // FIXME lack of resiliency, active rules index is corrupted if rule index fails
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java]: Code Smell Code Smell: Take the required action to fix the issue indicated by this comment. ([squid:S1134](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsb%2FSONAR-10052%2Fqprofiles&issues=AWE-qvQxmckxG3vIth7-&open=AWE-qvQxmckxG3vIth7-)

[ISSUE_KEY]: AWE-qvQxmckxG3vIth7-

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java [hunk 8]
```diff
@@ -59,37 +60,37 @@
...
+  public QProfileExporters(DbClient dbClient, RuleFinder ruleFinder, QProfileRules qProfileRules, ProfileImporter[] importers) {
+    this(dbClient, ruleFinder, qProfileRules, new ProfileExporter[0], importers);
   }
 
   /**
    * Used by Pico if no {@link ProfileImporter} nor {@link ProfileExporter} is found
    */
-  public QProfileExporters(DbClient dbClient, RuleFinder ruleFinder, RuleActivator ruleActivator) {
-    this(dbClient, ruleFinder, ruleActivator, new ProfileExporter[0], new ProfileImporter[0]);
+  public QProfileExporters(DbClient dbClient, RuleFinder ruleFinder, QProfileRules qProfileRules) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java]: Code Smell Code Smell: Remove this use of "RuleFinder"; it is deprecated. ([squid:CallToDeprecatedMethod](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsb%2FSONAR-10052%2Fqprofiles&issues=AWE-qvPNmckxG3vIth71&open=AWE-qvPNmckxG3vIth71)

[ISSUE_KEY]: AWE-qvPNmckxG3vIth71

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java [hunk 9]
```diff
@@ -59,37 +60,37 @@
...
     this.exporters = exporters;
     this.importers = importers;
   }
 
   /**
    * Used by Pico if no {@link ProfileImporter} is found
    */
-  public QProfileExporters(DbClient dbClient, RuleFinder ruleFinder, RuleActivator ruleActivator, ProfileExporter[] exporters) {
-    this(dbClient, ruleFinder, ruleActivator, exporters, new ProfileImporter[0]);
+  public QProfileExporters(DbClient dbClient, RuleFinder ruleFinder, QProfileRules qProfileRules, ProfileExporter[] exporters) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java]: Code Smell Code Smell: Remove this use of "RuleFinder"; it is deprecated. ([squid:CallToDeprecatedMethod](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsb%2FSONAR-10052%2Fqprofiles&issues=AWE-qvPNmckxG3vIth7z&open=AWE-qvPNmckxG3vIth7z)

[ISSUE_KEY]: AWE-qvPNmckxG3vIth7z

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java [hunk 10]
```diff
@@ -59,37 +60,37 @@
...
+    this(dbClient, ruleFinder, qProfileRules, exporters, new ProfileImporter[0]);
   }
 
   /**
    * Used by Pico if no {@link ProfileExporter} is found
    */
-  public QProfileExporters(DbClient dbClient, RuleFinder ruleFinder, RuleActivator ruleActivator, ProfileImporter[] importers) {
-    this(dbClient, ruleFinder, ruleActivator, new ProfileExporter[0], importers);
+  public QProfileExporters(DbClient dbClient, RuleFinder ruleFinder, QProfileRules qProfileRules, ProfileImporter[] importers) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java]: Code Smell Code Smell: Remove this use of "RuleFinder"; it is deprecated. ([squid:CallToDeprecatedMethod](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsb%2FSONAR-10052%2Fqprofiles&issues=AWE-qvPNmckxG3vIth70&open=AWE-qvPNmckxG3vIth70)

[ISSUE_KEY]: AWE-qvPNmckxG3vIth70

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java [hunk 11]
```diff
@@ -290,116 +272,74 @@ private ActiveRuleDto doInsert(ActiveRuleChange change, RuleActivatorContext con
     dao.insert(dbSession, activeRule);
     for (Map.Entry<String, String> param : change.getParameters().entrySet()) {
       if (param.getValue() != null) {
-        ActiveRuleParamDto paramDto = ActiveRuleParamDto.createFor(context.ruleParamsByKeys().get(param.getKey()));
+        ActiveRuleParamDto paramDto = ActiveRuleParamDto.createFor(rule.getParam(param.getKey()));
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java]: Bug Bug: "createFor" is marked "javax.annotation.ParametersAreNonnullByDefault" but parameter 1 could be null. ([squid:S2637](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2637))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsb%2FSONAR-10052%2Fqprofiles&issues=AWE-qvP5mckxG3vIth74&open=AWE-qvP5mckxG3vIth74)

[ISSUE_KEY]: AWE-qvP5mckxG3vIth74

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java [hunk 12]
```diff
@@ -176,87 +165,78 @@ private void updateProfileDates(DbSession dbSession, RuleActivatorContext contex
    * <p/>
    * On custom rules, it's always rule parameters that are used
    */
-  private void applySeverityAndParamToChange(RuleActivation request, RuleActivatorContext context, ActiveRuleChange change) {
+  private void applySeverityAndParamToChange(RuleActivation request, RuleActivationContext context, ActiveRuleChange change) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 27 to the 15 allowed. ([squid:S3776](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsb%2FSONAR-10052%2Fqprofiles&issues=AWE-qvP5mckxG3vIth79&open=AWE-qvP5mckxG3vIth79)

[ISSUE_KEY]: AWE-qvP5mckxG3vIth79

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/BuiltInQProfileUpdateImpl.java [hunk 13]
```diff
@@ -45,22 +47,31 @@ public BuiltInQProfileUpdateImpl(DbClient dbClient, RuleActivator ruleActivator,
...
-      changes.addAll(ruleActivator.activateOnBuiltInRulesProfile(dbSession, activation, ruleProfile));
-    });
+      activations.add(activation);
+      ruleKeys.add(activation.getRuleKey());
+      deactivatedKeys.remove(activation.getRuleKey());
+    }
+
+    RuleActivationContext context = ruleActivator.createContextForBuiltInProfile(dbSession, rulesProfile, ruleKeys);
+
+    List<ActiveRuleChange> changes = new ArrayList<>();
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/BuiltInQProfileUpdateImpl.java]: old school! you can write this with a flatmap :D

Comment (sns-seb) [reply server/sonar-server/src/main/java/org/sonar/server/qualityprofile/BuiltInQProfileUpdateImpl.java]: apparently, you could use flatmap in several locations in this commit

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileRulesImpl.java [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+    return changes;
+  }
+
+  private static void verifyNotBuiltIn(QProfileDto profile) {
+    checkArgument(!profile.isBuiltIn(), "The built-in profile %s is read-only and can't be updated", profile.getName());
+  }
+
+  private BulkChangeResult doBulk(DbSession dbSession, QProfileDto profile, RuleQuery ruleQuery, BiFunction<RuleActivationContext, RuleKey, List<ActiveRuleChange>> fn) {
+    BulkChangeResult result = new BulkChangeResult();
+    Collection<RuleKey> ruleKeys = Sets.newHashSet(ruleIndex.searchAll(ruleQuery));
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileRulesImpl.java]: why make a copy of the result here?
I didn't find any place where this was mutated (would be bad anyway). Any other reason?

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileTreeImpl.java [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+  private List<ActiveRuleChange> setParent(DbSession dbSession, QProfileDto profile, QProfileDto parent) {
+    checkRequest(parent.getLanguage().equals(profile.getLanguage()), "Cannot set the profile '%s' as the parent of profile '%s' since their languages differ ('%s' != '%s')",
+      parent.getKee(), profile.getKee(), parent.getLanguage(), profile.getLanguage());
+
+    List<ActiveRuleChange> changes = new ArrayList<>();
+    if (parent.getKee().equals(profile.getParentKee())) {
+      return changes;
+    }
+
+    checkRequest(!isDescendant(dbSession, profile, parent), "Descendant profile '%s' can not be selected as parent of '%s'", parent.getKee(), profile.getKee());
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileTreeImpl.java]: I didn't find a UT for this check

**Comment by sns-seb** [comment]
I'm concerned with the limit to 1000 rules, though

File: server/sonar-db-dao/src/main/java/org/sonar/db/qualityprofile/ActiveRuleDao.java [hunk 16]
```diff
@@ -76,18 +81,28 @@
     return mapper(dbSession).selectByRuleProfileUuid(ruleProfileDto.getKee());
   }
 
+  public Collection<ActiveRuleDto> selectByRulesAndRuleProfileUuids(DbSession dbSession, Collection<RuleDefinitionDto> rules, Collection<String> ruleProfileUuids) {
+    if (rules.isEmpty()) {
+      return emptyList();
+    }
+    checkArgument(rules.size() < PARTITION_SIZE_FOR_ORACLE,
```
[the remark this answers is not in the document]
Comment (simonbrandhof) [reply server/sonar-db-dao/src/main/java/org/sonar/db/qualityprofile/ActiveRuleDao.java]: The scope of method calls is always restricted to:
- a single built-in profile. Indeed profiles are registered at startup one by one. 
- a single user profile
So in both cases scope is restricted to a single profile, so by transitivity to a single language. The risk to have more than 1'000 rules in the same profile is quite low.

[the remark this answers is not in the document]
Comment (sns-seb) [reply server/sonar-db-dao/src/main/java/org/sonar/db/qualityprofile/ActiveRuleDao.java]: understood, thanks for the explanation

File: server/sonar-db-dao/src/main/java/org/sonar/db/qualityprofile/ActiveRuleDao.java [hunk 17]
```diff
@@ -76,18 +81,28 @@
     return mapper(dbSession).selectByRuleProfileUuid(ruleProfileDto.getKee());
   }
 
+  public Collection<ActiveRuleDto> selectByRulesAndRuleProfileUuids(DbSession dbSession, Collection<RuleDefinitionDto> rules, Collection<String> ruleProfileUuids) {
+    if (rules.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (simonbrandhof) [reply server/sonar-db-dao/src/main/java/org/sonar/db/qualityprofile/ActiveRuleDao.java]: Circuit breaker for `ruleProfileUuids` is implemented in `executeLargeInputs()`. However you're right for improving readability. The condition could be completed here.

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/BuiltInQProfileUpdateImpl.java (hunk 13 above)
Comment (simonbrandhof) [reply server/sonar-server/src/main/java/org/sonar/server/qualityprofile/BuiltInQProfileUpdateImpl.java]: I will check, thanks

File: server/sonar-db-dao/src/main/java/org/sonar/db/qualityprofile/ActiveRuleDao.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (simonbrandhof) [reply server/sonar-db-dao/src/main/java/org/sonar/db/qualityprofile/ActiveRuleDao.java]: Supporting more than 1'000 rules is trivial. Let's do it.

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileRulesImpl.java (hunk 14 above)
Comment (simonbrandhof) [reply server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileRulesImpl.java]: It enforces the unicity of the `Iterator` returned by `ruleIndex.searchAll(ruleQuery)`. Do you think it should be dropped?

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileTreeImpl.java (hunk 15 above)
Comment (simonbrandhof) [reply server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileTreeImpl.java]: See `set_grandchild_as_parent_fails()`. Note that I didn't refactor these tests :-)

File: server/sonar-db-dao/src/test/java/org/sonar/db/qualityprofile/ActiveRuleDaoTest.java [hunk 18]
```diff
@@ -175,6 +175,42 @@ public void selectByRuleProfileUuid() {
     assertThat(underTest.selectByProfile(dbSession, profile2)).isEmpty();
   }
 
+  @Test
+  public void selectByRulesAndRuleProfileUuids() {
```
[the remark this answers is not in the document]
Comment (simonbrandhof) [reply server/sonar-db-dao/src/test/java/org/sonar/db/qualityprofile/ActiveRuleDaoTest.java]: Test not required anymore, restriction on number of requested rules is dropped.

**Pull request merged**